File: 알고리즘 문제풀이/백준/효율적인 화폐 구성.py
import sys
INF = 10001
n,m = map(int,sys.stdin.readline().split(" "))
coins,dp=[],[INF]*(m+1)
dp[0]=0
for i in range(n):
    coins.append(int(sys.stdin.readline()))

#if dp[i-k] exist -> dp[i]=min(dp[i],dp[i-k]+1)
for coin in coins:
    for j in range(coin,m+1):
        if dp[j-coin]!=INF:
            dp[j] = min(dp[j],dp[j-coin]+1)
if dp[m]==INF:
    print(-1)
else:      
    print(dp[m])

File: 알고리즘 문제풀이/Algorithm/dynamic/효율적인화폐구성.js
const solution = (n, m) => {
  const dp = new Array(m + 1).fill(10001);

  dp[i] = 0;
  for (let i = 0; i < n.length; i++) {
    dp[n[i]] = 1;
  }

  for (let i = 2; i <= m; i++) {
    for (let j = 0; j < n.length; j++) {
      const coin = n[j];

      if (i - coin >= 0) {
        dp[i] = Math.min(dp[i], dp[i - coin] + 1);
      }
    }
  }

  /*
  for(let i=0;i<n.length;i++) {
    for(let j=n[i],i<=m;j++) {
        dp[j] = Math.min(dp[j],dp[j-n[i]]+1);
    }
  }
*/

  return dp[m] === 10001 ? -1 : dp[m];
};

console.log(solution([2, 3], 15));
console.log(solution([3, 5, 7], 4));
